File: mp1/puzzle.c
#include <math.h>
#include "puzzle.h"

int nodenum = 0;

int main(int argc, char *argv[])
{
  fringenode *head = NULL;
  fringenode *tmpnode;
  fringenode *nnode;
  int dim;
  int dimsqr;
  time_t start_time, end_time;
  int *istate = NULL;
  int found = 0;
  char *moves;
  int dist;
  int i;

  // read input state
  dim = read_state_from_stdin(&istate);
  if(dim == 0)
  {
    printf("Error: invalid input\n");
    return 0;
  }

if(PDEBUG) { printf("*after read\n"); print_state(istate, dim); printf("\n"); }

  // initialize head and state dimension
  dimsqr = dim * dim;
  head = create_node(istate, dim);
  nnode = head;

  moves = (char *)malloc(4 * sizeof(char));

  // time it!
  start_time = time(NULL);

  // is initial state a solution?
  dist = h_man_dist(nnode->state, dim);
  if(dist == 0)
  {
if(PDEBUG) printf("- init state is goal state\n");
    found = 1;
  }

  // A* for the solution
  while(!found)
  {
    if(head == NULL)  // empty list
    {
if(PDEBUG) printf("- head of list is empty at start\n");
      break;
    }

    nnode = head;
    // get possible moves from current state
    get_moves(moves, nnode, dim);

    // explore all moves
    i = 0;
    while(i < 4 && moves[i] != 'x')
    {
if(nodenum > 10) exit(0);
      // copy parent state to istate
      memcpy(istate, nnode->state, dimsqr * sizeof(int));

      // generate child state
      update_state(istate, moves[i], dim);

      // create child node and link with parent (prev_node)
      tmpnode = create_node(istate, dim);
      tmpnode->prev_node = nnode;
      tmpnode->f = nnode->f + 1;
      nnode->childs += 1;
if(PDEBUG) printf("- parent now has %d childs\n", nnode->childs);
if(PDEBUG) { printf("*in while child create\n"); print_state(tmpnode->state, dim); }

      // generate dist and goal check
      dist = h_man_dist(istate, dim);
      tmpnode->g = tmpnode->f + dist;
      if(dist == 0)
      {
        found = 1;
        continue;
      }

      // insert child into list (next_node)
      insert_node(tmpnode, &head);
if(PDEBUG) print_list_order(head, 2);

      ++i;
    }

  }

  end_time = time(NULL);

  if(found)
  {
    print_moves(nnode, 1);
  }
  printf("%d seconds\n", end_time - start_time);

  return 0;
}

/*
 * create a new node using an initial state
 */
fringenode *create_node(int *istate, int dim)
{
  int dimsqr = dim * dim;
  fringenode *nnode;

  nnode = (fringenode *)malloc(sizeof(fringenode));
  nnode->state = (int *)malloc(dimsqr * sizeof(int));
  memcpy(nnode->state, istate, dimsqr * sizeof(int));

  nnode->next_node = NULL;
  nnode->prev_node = NULL;
  nnode->f = 0;
  nnode->g = dimsqr;  // an arbitrary big number
  nnode->childs = 0;
  nnode->lastmove = 'x';  // x in initial state

nnode->nodenum = nodenum++;

  return nnode;
}

/*
 * fill the passed in state buffer with the initial state
 */
int read_state_from_stdin(int **state)
{
  int readdim = MAXDIM * 10;
  int instate[MAXDIM * MAXDIM];
  char userin[readdim];
  char *nexttok;
  int ldim = 0;
  int done = 0;
  int *lstate;

  while(!done)
  {
    if(fgets(userin, readdim, stdin) == NULL) // EOF
    {
      done = 1;
      continue;
    }

    userin[strlen(userin) - 1] = '\0';  // trim for strcmp

    nexttok = strtok(userin, " ");
    while(nexttok != NULL)
    {
      if(strcmp(nexttok, "X") != 0)
        // decrement number so it matches state index
        instate[ldim++] = atoi(nexttok) - 1;
      else
        instate[ldim++] = -1;

      nexttok = strtok(NULL, " ");
    }
  }

if(PDEBUG) printf("read: %d, dim: %d\n", ldim,
(int)sqrt(ldim));

  lstate = (int *)malloc(sizeof(int) * ldim);
  memcpy(lstate, instate, sizeof(int) * ldim);
  *state = lstate;

if(PDEBUG) printf(" state: %s, dim: %d\n",
(state == NULL) ? "NULL" : "NOT NULL", (int)sqrt(ldim));

  return (int)sqrt(ldim);
}

/*
 * recurisvely print out moves needed to get to node
 */
void print_moves(fringenode *fn, int terminal)
{
  if(fn->prev_node != NULL)
  {
    print_moves(fn->prev_node, 0);
  }

  if(fn->lastmove != 'x') // don't print init state move
    printf("%c", fn->lastmove);

  if(terminal)
    printf("\n");

}

/*
 * print an nxn grid of the state
 */
void print_state(int *state, int dim)
{
  int i;
  int dimsqr = dim * dim;

  for(i = 0; i < dimsqr; ++i)
  {
    if(state[i] == -1)
      printf("X ");
    else
      printf("%d ", state[i] + 1);

    if((i + 1) % 3 == 0)
      printf("\n");
  }

}

/*
 * change the state based on the given move
 */
void update_state(int *state, char move, int dim)
{
  int tmpidx, tmpcell, xcell;

  // find location of empty space (-1)
  xcell = 0;
  while(state[xcell] != -1)
    ++xcell;

  // find cell to switch
  switch(move)
  {
    case 'u':
      tmpidx = xcell - dim;
      break;
    case 'd':
      tmpidx = xcell + dim;
      break;
    case 'l':
      tmpidx = xcell - 1;
      break;
    case 'r':
      tmpidx = xcell + 1;
      break;
    default:
      printf("Error: unknown move in update_state\n");
  }

  // switch cells
  tmpcell = state[tmpidx];
  state[tmpidx] = state[xcell];
  state[xcell] = tmpcell;

}

/*
 * Manhatten distance heuristic function
 */
int h_man_dist(int *state, int dim)
{
  int dist = 0;
  int sr, sc, dr, dc; // source/dest row/col
  int tmpcell;
  int i;
  int dimsqr = dim * dim;

  for(i = 0; i < dimsqr; ++i)
  {
    if(i != state[i]) // cell out of place, get distance
    {
      tmpcell = state[i];
      if(tmpcell == -1)
        tmpcell = 8;

      sr = i / dim;
      sc = i % dim;
      dr = tmpcell / dim;
      dc = tmpcell % dim;

      if((dc = dc - sc) < 0)
        dc = dc * -1;
      if((dr = dr - sr) < 0)
        dr = dr * -1;
      
      dist += (dc + dr);
    }
  }

  return dist;
}

/*
 * determine all legal moves of node's state
 */
void get_moves(char *moves, fringenode *fn, int dim)
{
  int row, col, i;
/*
  for(i = 0; i < 4; ++i)
  {
    moves[i] = 'x';
  }
*/
  for(row = 0; row < 4; ++row)
  {
    for(col = 0; col < 4; ++col)
    {
      i = (row * dim) + col;
      if(fn->state[i] == -1)
        break;
    }

    if(fn->state[i] == -1)
      break;
  }

  i = 0;
  if(row != 0 && fn->lastmove != 'd')
  {
    moves[i++] = 'u';
if(PDEBUG) printf("- get_moves adding 'u'\n");
  }
  if(row != dim - 1 && fn->lastmove != 'u')
  {
    moves[i++] = 'd';
if(PDEBUG) printf("- get_moves adding 'd'\n");
  }
  if(col != 0 && fn->lastmove != 'r')
  {
    moves[i++] = 'l';
if(PDEBUG) printf("- get_moves adding 'l'\n");
  }
  if(col != dim - 1 && fn->lastmove != 'l')
  {
    moves[i++] = 'r';
if(PDEBUG) printf("- get_moves adding 'r'\n");
  }

  if(i < 4)
    moves[i] = 'x';

}

/* 
 * insert n node starting at h based on g value
 */
void insert_node(fringenode *n, fringenode **h)
{
  fringenode *tn; // test (next) node
  fringenode *cn = *h; // current node
int testcnt = 0;
if(PDEBUG) printf("-insert_node inserting node %d\n", n->nodenum);

  if(n->g < (*h)->g) // new node is better than head node
  {
    n->next_node = *h;
    *h = n;

if(PDEBUG) printf("-insert_node: new(%d) better than head(%d)\n",n->g,(*h)->g);
    return;
  }

  while(cn->next_node != NULL)
  {
testcnt++;
    tn = cn->next_node;
    if(n->g < tn->g)  // place new node before next node
      n->next_node = tn;
      tn->next_node = n;

if(PDEBUG) printf("-insert_node: new(%d) better than %d(%d)\n",n->g,testcnt,cn->g);
      return;
  }

  // place new node at the end
if(PDEBUG) printf("-insert_node: inserting at end\n");
  cn->next_node = n;

}

void print_list_order(fringenode *h, int count)
{
  int i = 0;
  fringenode *cn = h;

  printf("-list order: ");
  while(i < count && cn != NULL)
  {
    printf("%d ", cn->nodenum);
    cn = cn->next_node;
  }

  printf("\n");
}

File: mp1/python/puzzle.py
import heapq
import math
import sys
import timeit

##### class PriorityQueue #####

class PriorityQueue:
  def __init__(self):
    self._queue = []
    self._index = 0

  def push(self, item, priority):
    heapq.heappush(self._queue, (priority, self._index, item))
    self._index += 1

  def pop(self):
    if self._index == 0:
      return 0
    r = heapq.heappop(self._queue)[-1]
    self._index -= 1
    return r

##### class Anode #####

class Anode:
  def __init__(self):
    self.state = []
    self.dim = 0
    self.f = 0
    self.g = 99999
    self.h = 99999
    self.lastmove = 'x'
    self.parent = 0
    self.next_node = 0

  def SetState(self, initState):
    self.state = initState
    self.dim = int(math.sqrt(len(initState)))

  def ManDistFromGoal(self):
    dist = 0
    for i in range(len(self.state)):
      if self.state[i] != i:
        tmpcell = self.state[i]
        if tmpcell == -1:
          tmpcell = 8
        sr = i / self.dim
        sc = i % self.dim
        dr = tmpcell / self.dim
        dc = tmpcell % self.dim

        dist += int(math.fabs(dc - sc) + math.fabs(dr - sr))

    return dist

  def Man2DistFromGoal(self):
    dist = 0
    for i in range(len(self.state)):
      if self.state[i] != i:
        tmpcell = self.state[i]
        if tmpcell == -1:
          tmpcell = 8
        sr = i / self.dim
        sc = i % self.dim
        dr = tmpcell / self.dim
        dc = tmpcell % self.dim

        dist += int(math.fabs(dc - sc) + math.fabs(dr - sr))

        if dist != 1:
          dist = dist * 2

    return dist

  def EucDistFromGoal(self):
    dist = 0
    for i in range(len(self.state)):
      if self.state[i] != i:
        tmpcell = self.state[i]
        if tmpcell == -1:
          tmpcell = 8
        sr = i / self.dim
        sc = i % self.dim
        dr = tmpcell / self.dim
        dc = tmpcell % self.dim

        #dist += int(math.fabs(dc - sc) + math.fabs(dr - sr))
        dist += math.sqrt(math.pow((dc-sc), 2) + math.pow((dr-sr), 2))

    return dist

  def GenerateMoves(self):
    moves = []
    xcell = self.state.index(-1)
    row = xcell / self.dim
    col = xcell % self.dim

    if row != 0:
      moves += ['u']
    if row != (self.dim - 1):
      moves += ['d']
    if col != 0:
      moves += ['l']
    if col != (self.dim - 1):
      moves += ['r']

    if self.lastmove == 'r' and 'l' in moves:
      moves.remove('l')
    if self.lastmove == 'l' and 'r' in moves:
      moves.remove('r')
    if self.lastmove == 'u' and 'd' in moves:
      moves.remove('d')
    if self.lastmove == 'd' and 'u' in moves:
      moves.remove('u')

    return moves

  def GenerateNextState(self):
    xidx = self.state.index(-1)
    swapidx = -1

    if self.lastmove == 'u':
      swapidx = xidx - self.dim
    elif self.lastmove == 'd':
      swapidx = xidx + self.dim
    elif self.lastmove == 'l':
      swapidx = xidx - 1
    else:
      swapidx = xidx + 1

    tmpcell = self.state[xidx]
    self.state[xidx] = self.state[swapidx]
    self.state[swapidx] = tmpcell

  def PrintMoves(self, terminate):
    if self.lastmove == 'x':  # root node
      return

    self.parent.PrintMoves(False)

    sys.stdout.write(self.lastmove)
    if terminate == True:
      sys.stdout.write('\n')

  def PrintState(self):
    for i in range(0, len(self.state), self.dim):
      #print "%d %d %d" % tuple(self.state[i:i+self.dim])
      print ' '.join(str(e) for e in self.state[i:i+self.dim])

  def ExportState(self):
    s = ''
    a = self.state[:]
    a = map(lambda x : x + 1, a)
    xloc = a.index(0)
    a = map(str, a)
    a[xloc] = 'X'
    for i in range(0, len(a), self.dim):
      #s = s + "%s %s %s\n" % tuple(a[i:i+self.dim])
      s = s + ' '.join(a[i:i+self.dim]) + '\n'

    #s = s[:-1]

    return s

##### Utility functions #####

def readInput():
  tmpState = []
  for line in sys.stdin:
    tmpState += line.split()

  retState = []
  for i in range(len(tmpState)):
    if tmpState[i] == 'X':
      retState.append(-1)
    else:
      retState.append(int(tmpState[i]) - 1)

  return retState

##### Main function #####

def Main():
  TIMELIMIT = 1800  # 30 minutes
  PDEBUG = True
  moditer = 0
  USEEUCDIST = False
  USEMANDIST = True
  explored = {}
  found = False
  q = PriorityQueue()

  initState = readInput()

  # start timer
  start_time = timeit.default_timer()

  # wrap init state in a node, set nnode
  nnode = Anode()
  nnode.SetState(initState)
  if USEEUCDIST:
    dist = nnode.EucDistFromGoal()
  elif USEMANDIST:
    dist = nnode.ManDistFromGoal()
  else:
    dist = nnode.Man2DistFromGoal()
  nnode.h = dist
  if PDEBUG:
    print "-pushing node with g=%d" % (nnode.g)
  q.push(nnode, nnode.g)

  # A* loop
  while found == False:

    moditer = (moditer + 1) % 5000
    if moditer == 0:
      if timeit.default_timer() - start_time > TIMELIMIT:
        if PDEBUG:
          print "Time limit exceeded"
        break
      if PDEBUG:
        print "%d explored states in memory" % (len(q._queue))

    nnode = q.pop()
    if nnode == 0: # empty list
      break

    # is new state solution?
    if nnode.h == 0.0:
      found = True
      continue

    # generate valid moves from current state
    moves = nnode.GenerateMoves()

    # explore all moves
    for move in moves:

      # create copy slice of parent state
      childState = nnode.state[:]

      # wrap state and move in child class, set parent, f
      cnode = Anode()
      cnode.SetState(childState)
      cnode.lastmove = move
      cnode.parent = nnode
      cnode.f = nnode.f + 1

      # generate next state based on move
      cnode.GenerateNextState()
      ckey = ''.join(map(str, cnode.state))
      if explored.has_key(ckey):
        continue
      explored[ckey] = 1

      # insert new state into list
      if USEEUCDIST:
        dist = nnode.EucDistFromGoal()
      elif USEMANDIST:
        dist = nnode.ManDistFromGoal()
      else:
        dist = nnode.Man2DistFromGoal()
      cnode.h = dist
      cnode.g = cnode.f + dist
      if PDEBUG:
        print "-pushing node with g=%d" % (cnode.g)
      q.push(cnode, cnode.g)

  # stop timer
  stop_time = timeit.default_timer()

  # print solution if found
  if found == True:
    nnode.PrintMoves(True)
  elif PDEBUG:
    print "No solution found"

  # print time
  tot_time = stop_time - start_time

  #if tot_time < 1:
  #  print "%f seconds" % (tot_time)
  #else:
  print "%d seconds" % (int(tot_time))

  if PDEBUG:
    print "%d nodes explored" % (len(q._queue))

if __name__ == '__main__':
  Main()
